### PR TITLE
Move packages and runner to new crate

### PR DESCRIPTION
**Pull request by TimDiekmann**
This is part of [Better separation of crates](https://app.asana.com/0/1201707629991362/1201481007343142/f) (_internal_).

## 🌟 What is the purpose of this PR?

The `execution` crate includes the packages and runners/workers/workerpools.

## 🔗 Related links

- [Asana task](https://app.asana.com/0/1202055186269230/1202099498122333/f) _(internal)_
- [package system doc](https://app.asana.com/0/1199548034582004/1201644736959126/f) (_internal_)

## 🚫 Blocked by

- ~~Improve documentation on crate~~
- #493 

## 🔍 What does this change?

- **Add the execution crate:** (Excluding move-only commits or fixes for bugs introduced by this PR)
- [Rename SharedStore to SharedDatasets](https://github.com/hashintel/hash/pull/527/commits/012dde454fe62a3b67154b38ecdb56dce4e4513c)
- [Rename TaskSharedStore to SharedStore](https://github.com/hashintel/hash/pull/527/commits/90c955c8b636cb740581e0510683198f12df48e1)
- [Document SharedStore and accompanying API](https://github.com/hashintel/hash/pull/527/commits/bd6fcbfec3b20e137eee763d1fb9baa58ef4c9ec)
- [Make worker and worker_pool configs more clear](https://github.com/hashintel/hash/pull/527/commits/dddcba13129db81163f5ab73c4ac728a11aa285f)
- [Rename Worker to WorkerIndex](https://github.com/hashintel/hash/pull/527/commits/c1e014b30db319786597f3a70249faeb4e097077)
- [Fix output task verification changed in](https://github.com/hashintel/hash/pull/527/commits/b784c2de7337712d8750f386da6d624d0691276d) https://github.com/hashintel/hash/pull/361
- [Rename Task to PackageTask](https://github.com/hashintel/hash/pull/527/commits/e9eb3f5854539df863872e3532711ecea66aa2ab)
- [Merge GetTaskName, GetTaskArgs, and StoreAccessVerify into Task](https://github.com/hashintel/hash/pull/527/commits/dcefa6dcf609486118866e6bab033f1c94a5a62b)
- [Simplify Display implementation for package Names](https://github.com/hashintel/hash/pull/527/commits/74dfe08e264163247ce86dba4ec92a38ddd331be)
- [Rename Package and PackageCreator to include its type](https://github.com/hashintel/hash/pull/527/commits/f4c9265936e0aa3ea31f25b37b78740cbd34f047)
- [Rename package and creator traits to be more representative](https://github.com/hashintel/hash/pull/527/commits/25c6ae11de36039cf92806c2360464b4a341ef8f)
    - Especially, this enables linking to a common trait from documentation and makes the concept of packages clearer
- [Remove code duplication when getting the package ID](https://github.com/hashintel/hash/pull/527/commits/d2ffe26e9b08c37ea6a2ccd6b2a41713f109b695)
- [Introduce a PackageInitConfig](https://github.com/hashintel/hash/pull/527/commits/dbaa77c932daf2fc795e052b59643432fd914747)
- [Introduce a PackageCreatorConfig](https://github.com/hashintel/hash/pull/527/commits/1e59fd8be72598326d98019084c33db71c42ef99)
    - This makes the packages independent of global configurations
- [Rename Behavior to BehaviorFieldMap](https://github.com/hashintel/hash/pull/527/commits/7e185640697dcb1115a7415fafc4ce6d688dff06)
- [Rename SharedBehavior to Behavior](https://github.com/hashintel/hash/pull/527/commits/8fb58b718e662412e30a2c93606e069222fe3e20)
- [Make package name unambiguous](https://github.com/hashintel/hash/pull/527/commits/d3b58958d9cf782ccb3746a7838401f5a6db225e)
- [Add trait for Comms and ActiveTask (temporarily?)](https://github.com/hashintel/hash/pull/527/commits/a6387dc87c9486e9bcfcc948a141034befeec43b)
    - In order to move the PackageCreators, `Comms` needs to be known. To avoid moving the whole engine to `execution`, this was made a trait to be used when creating packages.
- [Remove Sync bound on Package](https://github.com/hashintel/hash/pull/527/commits/ff94bf805c9b1aa0d10393238a5aab8bc848afd4)
- [Make PackageComms generic over Comms](https://github.com/hashintel/hash/pull/527/commits/2b6ec8bcceeb43d1905ecda63b1b8676927215e2)
- [Make PackageCreators generic over Comms](https://github.com/hashintel/hash/pull/527/commits/f115e6153cd84b18310ca7fa4ea68c872a27ed33)
- [Fix package doc](https://github.com/hashintel/hash/pull/527/commits/f0e7d4dfeb54f5e0f01482252272527ec947a596)
- [Swap name of `Dataset` and `SharedDataset` to avoid confusion with `SharedDatasets`](https://github.com/hashintel/hash/pull/527/commits/8d5d4d319c5fd2ab2a7887b167bd1cb548c1259e)
- [Remove EngineConfig](https://github.com/hashintel/hash/pull/527/commits/8fde5bd05bcbaf9e32717c71de9aa20fca8e58f8)
- [Update paths in readme](https://github.com/hashintel/hash/pull/527/commits/82b5f9e36deb2f580411325a7affcb0603a4708d)
- [Add package system documentation](https://github.com/hashintel/hash/pull/527/commits/254524c944b71d7d0fd28bd8db35dd99f10f648f)
- [Add runner documentation](https://github.com/hashintel/hash/pull/527/commits/d0fd8571c3699667c89748ec4d5dfb779909f8d8)
- [Add task documentation](https://github.com/hashintel/hash/pull/527/commits/ca2f6369365d1c8fc619ced0a77287735b59acb2)

## 📜 Does this require a change to the docs?

- The readme was updated in 82b5f9e36deb2f580411325a7affcb0603a4708d
- The package system is described now in-code in 254524c944b71d7d0fd28bd8db35dd99f10f648f
- Documentation has been added for runners (d0fd8571c3699667c89748ec4d5dfb779909f8d8) and tasks (ca2f6369365d1c8fc619ced0a77287735b59acb2)

## ⚠️ Known issues

Packages and Tasks are supposed to be split. Tasks belong to the workers/runners, the packages should be stand-alone. Currently, they are coupled together quite tightly so this is out of scope for this PR.

## 🐾 Next steps

Add the worker and workerpools to this crate as well

## 🛡 What tests cover this?

The integration test suite covers this

## ❓ How to test this?

There are no real logic changes. Commits are either move-only commits, which should be reviewed by looking at the output of `cargo doc -p execution --no-deps`. The other commits are listed above, which in most cases are only renames or simple restructuring for making the moved structures independent.

## 📹 Demo

Run
```sh
cargo doc -p execution --no-deps
```

### CODE REVIEW

File: packages/engine/src/experiment/controller/controller.rs [hunk 1]
```diff
@@ -393,7 +393,7 @@ impl<P: OutputPersistenceCreatorRepr> ExperimentController<P> {
         exp_config: Arc<ExperimentConfig>,
         exp_base_config: Arc<ExperimentConfig>,
         env: Environment,
-        shared_store: Arc<SharedStore>,
+        shared_store: Arc<SharedDatasets>,
```
Comment (Alfred-Mountfield) [inline packages/engine/src/experiment/controller/controller.rs:403]: You also haven't renamed variables and such where this is used which is confusing. I'm thinking we might just want to revert this commit

File: packages/engine/lib/execution/src/task/shared_store.rs [hunk 2]
```diff
@@ -1,25 +1,27 @@
...
+    task::StateBatchDistribution,
+    worker_pool::{SplitConfig, WorkerAllocation, WorkerIndex},
+    Error, Result,
 };
 
-/// TODO: DOC, purpose and fields
+/// Holds proxies to access the state and information, if the context is readable.
+///
+/// If the `state` can accessed (i.e. `state` is not `SharedState::None`), it holds read-only,
+/// read-write, partially borrowed proxies to the state pools.
```
Comment (Alfred-Mountfield) [inline packages/engine/lib/execution/src/task/shared_store.rs:17]: Are there multiple StatePools? I think there's only one right?

File: packages/engine/lib/execution/src/worker_pool/config.rs [hunk 3]
```diff
@@ -1,21 +1,28 @@
...
+
 use crate::worker::WorkerConfig;
 
-pub type WorkerIndex = usize;
-pub type WorkerAllocation = Vec<Worker>;
+pub type WorkerAllocation = Vec<WorkerIndex>;
 
 #[derive(Clone, Copy, Debug, Eq, PartialEq, PartialOrd, Ord)]
-pub struct Worker(WorkerIndex);
+pub struct WorkerIndex(usize);
```
Comment (Alfred-Mountfield) [inline packages/engine/lib/execution/src/worker_pool/config.rs:8]: This is a great change, thank you :D

File: packages/engine/src/simulation/task/mod.rs [hunk 4]
```diff
@@ -103,7 +103,7 @@ use crate::simulation::{
 // Additionally we have TryInto<init::Task>, (and others)
 // implemented for this enum.
 #[derive(Clone, Debug)]
-pub enum Task {
+pub enum PackageTask {
```
Comment (Alfred-Mountfield) [inline packages/engine/src/simulation/task/mod.rs]: I'm not sure I like this change, (also the docs still need changing). I'm just not sure it has clearer meaning, I don't see these as being "Package" tasks. They're tasks, Packages happen to be able to spawn them.

Happy to hear the justification for this

**Review by Alfred-Mountfield — CHANGES_REQUESTED** [review]
Partial review, only got up to
> Merge `GetTaskName`, `GetTaskArgs`, and `StoreAccessVerify` into `Task`

Will continue review tomorrow. Feel free to update the branch in-between, will let you know when I'm reviewing again :)

File: packages/engine/Cargo.toml [hunk 5]
```diff
@@ -12,6 +12,8 @@ flatbuffers_gen = { path = "lib/flatbuffers_gen" }
 nano = { path = "lib/nano", default-features = false }
 memory = { path = "lib/memory" }
 stateful = { path = "lib/stateful" }
+simulation_structure = { path = "lib/simulation_structure" }
+execution = { path = "lib/execution" }
```
[the remark this answers is not in the document]
Comment (TimDiekmann) [reply packages/engine/Cargo.toml:16]: I knew I missed it somewhere 😄

File: packages/engine/lib/execution/src/task/shared_store.rs (hunk 2 above)
Comment (TimDiekmann) [reply packages/engine/lib/execution/src/task/shared_store.rs:17]: It's intentionally not saying `StatePools` but "state pools" and refers to the agent pool and message pool, thus the plural. Changed the docs to "partially borrowed proxies to agents and messages" for clarification.

File: packages/engine/src/simulation/task/mod.rs (hunk 4 above)
Comment (TimDiekmann) [reply packages/engine/src/simulation/task/mod.rs]: This was mainly done to avoid confusion between `enum Task` and `trait Task`. If we decide to revert the changes for `trait Task`, I'm happy to revert this as well.
Additionally, this was done because the variants are specifically for the different packages, so the tasks are very specific.

File: packages/engine/src/simulation/package/init/packages/mod.rs [hunk 6]
```diff
@@ -70,16 +68,24 @@ pub enum InitTask {
     PyInitTask(PyInitTask),
 }
 
-impl GetTaskName for InitTask {
-    fn get_task_name(&self) -> &'static str {
+impl Task for InitTask {
+    fn name(&self) -> &'static str {
```
[the remark this answers is not in the document]
Comment (TimDiekmann) [reply packages/engine/src/simulation/package/init/packages/mod.rs]: Just for clarification: Are you against `trait Task` or the way where the actual logic happens?

File: packages/engine/src/datastore/shared_store.rs [hunk 7]
```diff
@@ -5,16 +5,15 @@ use stateful::global::Dataset;
 
 use crate::{config::ExperimentConfig, datastore::error::Result, proto::ExperimentRunTrait};
 
-// TODO rename to something more self-explanatory
 /// This is an object we use to manage sharing access to data that's static across simulation runs
 /// within an experiment, such as datasets.
 #[derive(Clone)]
-pub struct SharedStore {
+pub struct SharedDatasets {
```
[the remark this answers is not in the document]
Comment (TimDiekmann) [reply packages/engine/src/datastore/shared_store.rs]: I reverted the naming: [Revert naming on shared stores](https://github.com/hashintel/hash/pull/527/commits/4be18fdc06017311a57e4e001ddc81292a74f7ab)

File: packages/engine/src/datastore/table/task_shared_store.rs [hunk 8]
```diff
@@ -15,12 +15,12 @@ use crate::{
 
 /// TODO: DOC, purpose and fields
 #[derive(Default, Debug)]
-pub struct TaskSharedStore {
+pub struct SharedStore {
```
[the remark this answers is not in the document]
Comment (TimDiekmann) [reply packages/engine/src/datastore/table/task_shared_store.rs]: I reverted the naming: [Revert naming on shared stores](https://github.com/hashintel/hash/pull/527/commits/4be18fdc06017311a57e4e001ddc81292a74f7ab)

File: packages/engine/lib/execution/src/task/shared_store.rs [hunk 9]
```diff
@@ -29,21 +31,29 @@ impl TaskSharedStore {
     }
 }
 
-/// TODO: DOC
+/// Partial write access to the state pools.
+///
+/// This only holds a subset of the state pools' batches specified by `group_indices`. Used in
```
[the remark this answers is not in the document]
Comment (TimDiekmann) [reply packages/engine/lib/execution/src/task/shared_store.rs]: Same as above: [Clarify documentation on `PartialStateWriteProxy`](https://github.com/hashintel/hash/pull/527/commits/4cbeeb6c1bf3c795b8e526b32d6f32705067c9e2)

File: packages/engine/src/simulation/package/init/packages/mod.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (TimDiekmann) [reply packages/engine/src/simulation/package/init/packages/mod.rs]: I moved the task logic to the concrete tasks: [Move task logic to concrete tasks](https://github.com/hashintel/hash/pull/527/commits/98a5b19db1f124a93c2aea0fff01cef89d509133)

File: packages/engine/src/simulation/package/state/packages/mod.rs [hunk 10]
```diff
@@ -68,18 +66,33 @@ pub enum StateTask {
...
+            Self::ExecuteBehaviorsTask(_) => "BehaviorExecution",
         }
     }
-}
 
-impl GetTaskArgs for StateTask {
     fn distribution(&self) -> TaskDistributionConfig {
         match self {
-            Self::ExecuteBehaviorsTask(inner) => inner.distribution(),
+            Self::ExecuteBehaviorsTask(_) => {
```
[the remark this answers is not in the document]
Comment (TimDiekmann) [reply packages/engine/src/simulation/package/state/packages/mod.rs]: I moved the task logic to the concrete tasks: [Move task logic to concrete tasks](https://github.com/hashintel/hash/pull/527/commits/98a5b19db1f124a93c2aea0fff01cef89d509133)

File: packages/engine/lib/execution/src/package/state/behavior_execution/behavior/field.rs [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+use crate::{
+    package::{
+        state::behavior_execution::{behavior::keys::BehaviorKeys, Behavior},
+        PackageInitConfig,
+    },
+    Error, Result,
+};
+
+#[derive(Clone)]
+pub struct BehaviorFieldMap {
```
Comment (Alfred-Mountfield) [inline packages/engine/lib/execution/src/package/state/behavior_execution/behavior/field.rs:14]: I don't have a good suggestion, but can you think of a better name for this? It's just a bit confusing when it **isn't** a `Map` and we literally have `BehaviorMap` below

File: packages/engine/lib/execution/src/package/comms.rs [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+    ) -> Result<C::ActiveTask> {
+        let task_name = task.name();
+
+        self.comms
+            .new_task(self.package_id, task, shared_store)
+            .instrument(tracing::debug_span!("Task", name = task_name))
+            .await
+    }
+
+    //     // TODO: UNUSED: Needs triage
```
Comment (Alfred-Mountfield) [inline packages/engine/lib/execution/src/package/comms.rs:58]: We should probably decide what to do with this in this PR.

The intention is that Packages should be able to create and remove agents as well.

I think considering the implementation is already there we should probably leave it in and just allow unused code in clippy until we add more packages

File: packages/engine/lib/execution/src/package/init/js_py/message.rs [hunk 13]
```diff
@@ -5,8 +5,11 @@ use stateful::agent::Agent;
 
 #[derive(Clone, Debug, Serialize, Deserialize)]
 pub enum JsPyInitTaskMessage {
+    #[serde(rename = "StartMessage")]
```
Comment (Alfred-Mountfield) [inline packages/engine/lib/execution/src/package/init/js_py/message.rs:8]: I think we might just want to change the code on the other end (i.e. in the package language impls) because this isn't going to be ergonomic to do on all package message types

File: packages/engine/lib/execution/src/lib.rs [hunk 14]
```diff
@@ -0,0 +1,18 @@
+//! # Workers, language runners and the hEngine Package System
```
Comment (Alfred-Mountfield) [inline packages/engine/lib/execution/src/lib.rs:1]: I don't think this is a good opening doc line to be honest. We want to describe what the _point_ of the crate is. 

(Don't just copy paste this because I'm not spending too much time thinking about phrasing) Something like
Handles what logic is applied, and when, during simulation as well as how that work is distributed and executed.

**Review by Alfred-Mountfield — CHANGES_REQUESTED** [review]

File: packages/engine/lib/execution/src/package/mod.rs [hunk 15]
```diff
@@ -0,0 +1,268 @@
+//! # The hEngine Package System
```
Comment (Alfred-Mountfield) [inline packages/engine/lib/execution/src/package/mod.rs:1]: Don't really know where to leave this comment so just adding it here. I'm a little worried about the folder structure. Previously there was some idea of a "Experiment Packages" and "Simulation Packages" and now we seem to just have "package".

Can you help explain? :)

File: packages/engine/lib/execution/src/task/shared_store.rs (hunk 2 above)
Comment (TimDiekmann) [reply packages/engine/lib/execution/src/task/shared_store.rs]: Fixed other occurrences I found: [Improve wording for state pools](https://github.com/hashintel/hash/pull/527/commits/1d4e1805d90532ed5b6bdfcffc5b8798322e8408)

File: packages/engine/lib/execution/src/package/mod.rs [hunk 16]
```diff
@@ -0,0 +1,268 @@
...
+    package_type::PackageType,
+    task::PackageTask,
+};
+use crate::Result;
+
+pub trait PackageCreator: Send + Sync {
+    /// A message sent to all workers before running any packages.
+    ///
+    /// This allows package creators to pass any kind of configuration from their Rust runtime to
+    /// their Language Runner counterpart all.
```
[the remark this answers is not in the document]
Comment (TimDiekmann) [reply packages/engine/lib/execution/src/package/mod.rs]: Huh 👀 
Thanks for providing the original docs: f8057ab66a7aa2a7596a8c2acc8be66715396880

File: packages/engine/lib/execution/src/package/mod.rs [hunk 17]
```diff
@@ -0,0 +1,268 @@
...
+
+// Generics for packages
+//
+// Note that generalization cannot be done for all package traits.
+// See https://github.com/rust-lang/rust/issues/20400 for why Init, Context, State and Output
+// package (and other respective) traits cannot be (sensibly) generalized into one even though they
+// are clearly disjoint.
+
+pub trait Package: Send {
+    /// A message sent to the workers before the package is running.
```
[the remark this answers is not in the document]
Comment (TimDiekmann) [reply packages/engine/lib/execution/src/package/mod.rs:249]: I already worried that this isn't clear, nice to be confirmed here 😄 
I included a few words to explicitly tell, that the messages are sent once per sim/exp. Also, I removed a redundant sentence from the `init_message` doc: fd6e8aeb7de374c935885587c7f3b2b804c695d4

Do you prefer **once per simulation** or *once per simulation* btw.?

File: packages/engine/lib/execution/src/package/init/task.rs [hunk 18]
```diff
@@ -16,12 +18,21 @@ pub enum InitTask {
 impl Task for InitTask {
     fn name(&self) -> &'static str {
         match self {
-            Self::JsInitTask(_) => "JsInit",
-            Self::PyInitTask(_) => "PyInit",
+            Self::JsInitTask(task) => task.name(),
+            Self::PyInitTask(task) => task.name(),
         }
     }
```
[the remark this answers is not in the document]
Comment (TimDiekmann) [reply packages/engine/lib/execution/src/package/init/task.rs]: Yes, too bad! 🙃 
I should create a crate for this.

File: packages/engine/lib/execution/src/package/state/behavior_execution/behavior/field.rs (hunk 11 above)
Comment (TimDiekmann) [reply packages/engine/lib/execution/src/package/state/behavior_execution/behavior/field.rs:14]: I had issues finding a name for this, too. I also find the term `Map` confusing but wasn't able to come up with a proper name.

Comment (Alfred-Mountfield) [reply packages/engine/lib/execution/src/package/state/behavior_execution/behavior/field.rs:14]: Maybe BehaviorAndFields? BehaviorWithFields?

Although we should probably just rename BehaviorKeys to BehaviorFields so maybe add a todo for that and leave the naming as is

File: packages/engine/lib/execution/src/package/comms.rs (hunk 12 above)
Comment (TimDiekmann) [reply packages/engine/lib/execution/src/package/comms.rs]: Yes, let's keep them: 1c02fb52eef8296de3ea52faa98adaf9b357ff1e

File: packages/engine/lib/execution/src/package/mod.rs [hunk 19]
```diff
@@ -221,11 +221,10 @@ pub trait PackageCreator: Send + Sync {
     /// A message sent to all workers before running any packages.
     ///
     /// This allows package creators to pass any kind of configuration from their Rust runtime to
-    /// their Language Runner counterpart all.
+    /// their Language Runner counterpart for the experiment.
     ///
     /// Compared to [`Package::start_message()`], the data returned with this method will be
-    /// available for all simulations. Also, this should not be implemented for packages but
-    /// their respective package creator.
+    /// available for all simulations and is sent **once per experiment**.
```
Comment (Alfred-Mountfield) [inline packages/engine/lib/execution/src/package/mod.rs:227]: I still think the wording (and naming of the methods) isn't clear. Lets try and rethink this from a high-level.

These methods are implemented by Packages (and their Creators) to define how they register themselves with the Language Runners, and to provide any initialisation data that they need to. The one on the `PackageCreator` is sent at the start of the experiment. (This is why it's on the `PackageCreator` because `Package`s are scoped to a Simulation but the Creators, i.e. factories, are unique)

The one on the `Package` is sent at the start of a simulation by a Package when it's created. This allows the Package to send initialisation data to the language runners.

Do you know where the messages actually go? As in what code they reach within the language runners, I'm assuming the package impls?

File: packages/engine/lib/execution/src/package/mod.rs (hunk 17 above)
[the remark this answers is not in the document]
Comment (TimDiekmann) [reply packages/engine/lib/execution/src/package/mod.rs:249]: I know, but it's important for the docs as well I think.

File: packages/engine/lib/execution/src/task/shared_store.rs [hunk 20]
```diff
@@ -41,17 +41,17 @@ pub struct PartialStateWriteProxy {
     pub state_proxy: StateWriteProxy,
 }
 
-/// Partial read access to the state pools.
+/// Partial read access to the agent and message pool.
```
Comment (Alfred-Mountfield) [inline packages/engine/lib/execution/src/task/shared_store.rs:44]: Tiny nit, sorry :D 

```suggestion
/// Partial read access to the agent and message pools.
```
or 
```suggestion
/// Partial read access to the agent pool and message pool.
```

(and on the other occurrences)

File: packages/engine/lib/execution/src/package/comms.rs [hunk 21]
```diff
@@ -70,12 +70,12 @@ impl<C: Comms> PackageComms<C> {
             .await
     }
 
-    #[allow(dead_code)]
+    // TODO: UNUSED: triage
```
Comment (Alfred-Mountfield) [inline packages/engine/lib/execution/src/package/comms.rs:73]: Why did you add these back? This doesn't need a triage, we've decided to keep it (we should add a comment saying why it's unused at the moment as _the current_ packages don't need to add/remove agents)

File: packages/engine/lib/execution/src/package/mod.rs (hunk 17 above)
[the remark this answers is not in the document]
Comment (Alfred-Mountfield) [reply packages/engine/lib/execution/src/package/mod.rs:249]: I've suggested we should rewrite the whole thing in the other comment thread, but in this specific example I'd probably go with 

> once per **simulation**

File: packages/engine/lib/execution/src/task/shared_store.rs (hunk 20 above)
Comment (TimDiekmann) [reply packages/engine/lib/execution/src/task/shared_store.rs]: I chose the latter to distinguish between singularity and plurality of pools.

File: packages/engine/lib/execution/src/package/comms.rs (hunk 21 above)
Comment (TimDiekmann) [reply packages/engine/lib/execution/src/package/comms.rs]: I added a description why they are unused and also added a doc-string: f1067a787f4e9440738fea6e8fdf080a3966922c

File: packages/engine/lib/execution/src/package/mod.rs (hunk 19 above)
Comment (TimDiekmann) [reply packages/engine/lib/execution/src/package/mod.rs]: The `init_message` method is only used for behavior execution and is used to create `Vec<BehaviorDescription>`:
```rust
pub struct BehaviorDescription {
    pub id: BehaviorId,
    pub name: String,
    pub short_names: Vec<String>,
    pub source: String,
    pub required_field_keys: Vec<String>,
    pub language: Language,
    pub dyn_access: bool,
}
```

It's encapsulated in `ExperimentInitRunnerMsgBase`, which contains general information about the experiment, which is passed to `spawn_worker`.

The `start_message` isn't used at all currently. The messages are passed with `NewSimulationRun` to all runners when a new simulation starts.
```rust
pub struct NewSimulationRun {
    pub span: Span,
    pub short_id: SimulationShortId,
    pub worker_allocation: Arc<WorkerAllocation>,
    pub packages: PackageMsgs,
    pub datastore: DatastoreSimulationPayload,
    pub globals: Arc<Globals>,
}
```

Any naming ideas? I think this high level overview should simplify the naming a little bit. The former is used like a `worker_init_message` and the latter is used as a `simulation_start_message`.

File: packages/engine/lib/execution/src/package/mod.rs (hunk 15 above)
Comment (TimDiekmann) [reply packages/engine/lib/execution/src/package/mod.rs:1]: Discussed off-GitHub: We will add experiment packages to this crate as well and make the distinction in the follow-up PR.

File: packages/engine/lib/execution/src/lib.rs (hunk 14 above)
Comment (TimDiekmann) [reply packages/engine/lib/execution/src/lib.rs]: I improved the crate level documentation, should be far better now. Also adjusted the docs on `TaskMessage` as it contained outdated information and weird formatting.

File: packages/engine/lib/execution/src/package/mod.rs (hunk 19 above)
Comment (Alfred-Mountfield) [reply packages/engine/lib/execution/src/package/mod.rs]: I think those suggestions are good. (Isn't that basically what they were before? 😁 )

File: packages/engine/src/simulation/task/mod.rs (hunk 4 above)
Comment (Alfred-Mountfield) [reply packages/engine/src/simulation/task/mod.rs]: I think I'm okay with it as is, at least it's an improvement on how it was before I think. We can always revisit if it doesn't seem to make sense / is confusing later on
